add documentation for filing package object (#1001)
bcgov/jag-file-submission: docs/README.md

# jag-file-submission

Generic File Submission API (to be used by the Family Law Act Application at first)

## Introduction

> Ready for integrating your application with bcgov efiling service and allow your users to efile document ?

Reviewer: ```suggestion
Ready to integrate your application with bcgov's eFiling service and allow your users to eFile documents?
```

> Ready for integrating your application with bcgov efiling service and allow your users to efile document ?

Reviewer 2: ```suggestion
Are you ready to integrate your application with the BC Gov E-filing Hub in order to allow your users to e-file documents?
```

Reviewer 2: iunno tho

> This guide will provide you with all information you need to get started.

Reviewer: ```suggestion
This guide will provide you with all the required information you need to get started.
```

### Request a profile

> Follow the following instructions in [Getting started - Onboarding](onboarding.md)

### Start Integrating

> Get an oauth token from our Keycloak Server using the following curl

Reviewer: ```suggestion
Get an oauth token from our Keycloak Server using the following curl command:
```

```bash

> curl --location --request POST 'url' \

Reviewer 2: ```suggestion
curl --location --request POST 'keycloak_url' \
```

--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=[your client id]' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'client_secret=[your client secret]'
```

> the secret should be kept safe and not exposed to the public

> Start by uploading document(s) to the **efiling hub** api

Reviewer: ```suggestion
Start by uploading document(s) to the **eFiling Hub** API:
```

```bash

> curl --location --request POST 'http://fla-nginx-proxy-qzaydf-dev.pathfinder.gov.bc.ca/api/submission/documents' \

Reviewer 2: is our reverse proxy URL meant to be public?

Author: removed

--header 'X-Transaction-Id: ca09e538-d34e-11ea-87d0-0242ac130003' \
--header 'Content-Type: multipart/form-data' \

> --header 'Authorization: Bearer [your bearer token]' \

Reviewer 2: ```suggestion
--header 'Authorization: Bearer [bearer_token]' \
```

--form 'files=test.pdf'
```

the response includes a submission id that you will use to generate a redirect url

```json
{
  "submissionId": "5e9492cf-e87e-48b5-ba55-0c198d8edde5",
  "received": 1
}
```

> Then generate a unique url to redirect the users to the **efiling hub**

Reviewer: ```suggestion
Then generate a unique url to redirect the users to the **eFiling Hub**
```

```bash
curl --location --request POST 'http://fla-nginx-proxy-qzaydf-dev.pathfinder.gov.bc.ca/api/submission/5e9492cf-e87e-48b5-ba55-0c198d8edde5/generateUrl' \
--header 'X-Transaction-Id: ca09e538-d34e-11ea-87d0-0242ac130003' \
--header 'Content-Type: application/json' \

> --header 'Authorization: Bearer [your bearer token]' \

Reviewer 2: ```suggestion
--header 'Authorization: Bearer [bearer_token]' \
```

--data-raw 'the json payload bellow'
```

payload:

```json
{
  "clientAppName": "string",
  "filingPackage": {
    "court": {
      "location": "string",
      "level": "P",
      "courtClass": "F",
      "division": "I",
      "fileNumber": "string"
    },
    "documents": [
      {
        "name": "string",
        "type": "ABP",
        "isAmendment": true,
        "isSupremeCourtScheduling": true,
        "data": {},
        "md5": "string"
      }
    ],
    "parties": [
      {
        "partyType": "IND",
        "roleType": "ABC",
        "firstName": "string",
        "middleName": "string",
        "lastName": "string"
      }
    ]
  },
  "navigationUrls": {
    "success": "string",
    "error": "string",
    "cancel": "string"
  }
}
```

### Generate URL Payload Details

#### navigation

The `navigation` object represents a list of possible returns to your application based on the status of the document e-filing.

> #### clientAppplication

Reviewer: ```suggestion
#### clientApplication
```

> The `clientAppplication` object represents how your application is labelled in **efiling hub**.

Reviewer: ```suggestion
The `clientApplication` object represents how your application is labelled in **efiling hub**.
```

#### filingPackage

> The `filingPackage` object represents the court information about the submited package along with the document(s) info.

Reviewer: ```suggestion
The `filingPackage` object represents the court information about the submitted package along with the document(s) info.
```

##### court

The `court` object represents the court details.

##### Properties

| name       | type   | description                                                                   |
| ---------- | ------ | ----------------------------------------------------------------------------- |
| location   | string | Court House where the package is submitted to                                 |
| level      | string | Court level, can be `Supreme` of `Provincial`                                 |
| courtClass | string | Court Classification see [Court Classification](data.md#Court-Classification) |
| division   | string | R for Criminal, I for Civil see [Court Division](data.md#Court-Division)      |
| fileNumber | string | The court file number                                                         |

##### Documents

> The `documents` array represents the previously uploaded document.

Reviewer: ```suggestion
The `documents` array represents the previously uploaded documents.
```

###### Properties

| name                     | type    | description                                                                                |
| ------------------------ | ------- | ------------------------------------------------------------------------------------------ |
| name                     | string  | the document name, must be the same name as uploaded previously                            |
| type                     | string  | the type of document see [Document Type Codes](data.md#Document-Type-Codes)                |
| isAmendment              | boolean | if the document is an amendment                                                            |

> | isSupremeCourtScheduling | boolean | is the document is directed to Supreme Court Scheduling                                    |

Reviewer: ```suggestion
| isSupremeCourtScheduling | boolean | if the document is directed to Supreme Court Scheduling                                    |
```

> | data                     | object  | A non define json object representing the form data of the document                        |

Reviewer: ```suggestion
| data                     | object  | A non defined json object representing the form data of the document                        |
```

| md5                      | string  | The md5 hash value of the document content, used to validate the integrity of the document |

##### parties

The `parties` array represents a list of parties for the submission

##### Properties

| name       | type   | description                                                                                     |
| ---------- | ------ | ----------------------------------------------------------------------------------------------- |
| partyType  | string | the party type, IND for individual or ORG for Organization see [Party Type](data.md#Party Type) |
| roleType   | string | the party role, see [Party Role](data.md#Party-Role)                                            |
| firstName  | string | the party first name                                                                            |
| middleName | string | the party middle name                                                                           |
| lastName   | string | the party last name                                                                             |

> 

Reviewer: as part of the generate url payload info we should also include a section about the validation.

1) must provide a court file number. If this is not provided, then parties with valid role type are mandatory

2) must provide valid document types

3) must provide a valid court level, class, and location combination

Author: I have added a required column in all the tables

### Api Documentation

> E-filing-api is documented using [openapi](http://editor.swagger.io/?url=https://raw.githubusercontent.com/bcgov/jag-file-submission/master/src/backend/efiling-api/jag-efiling-api.yaml) and we have a [postman collection](https://raw.githubusercontent.com/bcgov/jag-file-submission/master/src/backend/jag-efiling-api/src/test/jag-efiling-api.postman_collection.json) that you can use to test the e-filing-api

Reviewer: ```suggestion
eFiling API is documented using [openapi](http://editor.swagger.io/?url=https://raw.githubusercontent.com/bcgov/jag-file-submission/master/src/backend/efiling-api/jag-efiling-api.yaml) and we have a [postman collection](https://raw.githubusercontent.com/bcgov/jag-file-submission/master/src/backend/jag-efiling-api/src/test/jag-efiling-api.postman_collection.json) that you can use to test the eFiling API
```
